pkg/services/ngalert/schedule/schedule.go

```diff
@@ -487,10 +515,11 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 				}
 			}()
 		case <-grafanaCtx.Done():
-			// clean up the state only if the reason for stopping the evaluation loop is that the rule was deleted
-			if errors.Is(grafanaCtx.Err(), errRuleDeleted) {
-				clearState()
-			}
+			// The context deadline is 5 seconds which should be enough time for the state to be deleted from the database
+			// and the historian to record the state changes async. Do not cancel the context with defer as otherwise
+			// the context will be canceled before the historian can write any state changes.
+			ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
```

Comment:
@grobinson-grafana this cancel function discard is making the lint to fail, how should we proceed?

Reply:
I have approved PR https://github.com/grafana/grafana/pull/61858. Let's wait for it to be merged and implement it properly